### ISSUE TITLE
Doubts aboutTransfer learning

### ISSUE THREAD

**Issue by Juanjojr9** [ultralytics/yolov5#7596]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi @ramonhollands  @glenn-jocher 

Referring to this comment : 

Might be interesting to do a final step by unfreezing and training the complete netwerk again with differentiated learning rate. So complete training process would be (default method in fast.ai):
> 
> 1. Freeze the backbone
> 2. (optional reset the head weights)
> 3. Train the head for a while
> 4. Unfreeze the complete network
> 5. Train the complete network with lower learning rate for backbone.

I have the following question:
I would like to ask you about this experiment. First of all, what is the benefit of freezing only the backbone, training the network and then retraining the whole network? I've been searching but I can't find anything related to this. I would like to apply this technique, but I don't know what is the benefit of training it that way, instead of training the net only once with the backbone frozen. On the other hand, what are the disadvantages?

Based on your evidence, is it a correct way to apply transfer learning in this way? what results you obtained?

I hope you can help me. 
Best regards

**Comment by glenn-jocher** [comment]
@Juanjojr9 I wouldn't recommend freezing any layers. Freezing is just a way to get mediocre results faster, but in my experience you'll usually get best results training normally.

**Comment by ramonhollands** [comment]
Hi Juanjojr9,

The intuition behind this technique is that we probably don't want to touch the very first layers in the backbone too much because they are so general meaningfull for different kind of images / domains. E.g. edges in the very first layers and circles  in some later layers. But in the end we do want to finetune these layers for the different task.

This is even more applicable for custom heads put on an pretrained backbone. Like using a Resnet imagenet trained classification model and replacing the last layers for a 'bounding-box head'. The head will be 'randomly' initialized. (Which is not the case for a yolov5 trained model)

Unfortunately I did not experiment with it (yet).

Best regards,
Ramon

**Comment by Juanjojr9** [comment]
> @Juanjojr9No recomendaría congelar ninguna capa. La congelación es solo una forma de obtener resultados mediocres más rápido, pero en mi experiencia, generalmente obtendrá mejores resultados entrenando normalmente.

@glenn-jocher  So, you don't recommend using the learning transfer, do you? When would you use this technique?

Only train the whole network.

**Comment by Juanjojr9** [comment]
> Hi Juanjojr9,
> 
> The intuition behind this technique is that we probably don't want to touch the very first layers in the backbone too much because they are so general meaningfull for different kind of images / domains. E.g. edges in the very first layers and circles in some later layers. But in the end we do want to finetune these layers for the different task.
> 
> This is even more applicable for custom heads put on an pretrained backbone. Like using a Resnet imagenet trained classification model and replacing the last layers for a 'bounding-box head'. The head will be 'randomly' initialized. (Which is not the case for a yolov5 trained model)
> 
> Unfortunately I did not experiment with it (yet).
> 
> Best regards, Ramon

@ramonhollands Ok, I understand what you say in the first paragraph.
But, I still don't understand what benefit there could be in training the network with the first layers frozen and then re-training the network but this time in its entirety. Why would it be good to use this way?

**Comment by glenn-jocher** [comment]
@Juanjojr9 freezing layers might have use cases in resource constrained environments, i.e. training on an edge device, or when you want faster training, and are willing to sacrifice accuracy. If you have resources and want the best results I would not freeze any layers.

**Comment by Juanjojr9** [comment]
@glenn-jocher First of all, thank you very much for getting back to me so quickly, I appreciate it.

In my case, I have limited resources, so I prefer to sacrifice a little precision.

In your opinion, would you recommend retraining the whole net, after having trained the net with a frozen backbone? Would there be any advantages or positives to doing that?

**Comment by glenn-jocher** [comment]
@Juanjojr9 I don't know, freezing's not my specialty, but if you consider freezing the entire model on one end and training normally on the other I'm sure there's an infinite range of training strategies you can implement that will all lie on a curve of accuracy vs training time. It's up to you to experiment in that space.

**Comment by Juanjojr9** [comment]
@glenn-jocher Amm... ok, I understand. My idea was to do a training with the spine frozen, not the head, and another normal training without freezing anything.

That's why I wanted to know if you had done any experiments on this issue, before doing anything and be sure of what I'm doing, bad luck...

**Comment by glenn-jocher** [comment]
@Juanjojr9 It's always good to experiment. You might be interested in trying out different training strategies and comparing the results. Good luck with your experiments!
